app/helpers/automated_tests_helper.rb

```diff
@@ -480,6 +486,435 @@ def parse_test_output(repo_assignment_dir, assignment, logfile, data)
+  def self.perform(grouping_id, call_on)
+    # Pick a server, launch the Test Runner and wait for the result
+    # Then store the result into the database
+
+    #@submission = Submission.find(submission_id)
+    @grouping = Grouping.find(grouping_id)
+    @assignment = @grouping.assignment
+    @group = @grouping.group
+    @repo_dir = File.join(MarkusConfigurator.markus_config_automated_tests_repository, @group.repo_name)
+
+    @list_of_servers = MarkusConfigurator.markus_ate_test_server_hosts.split(' ')
+
+    while true
+      @test_server_id = choose_test_server()
+      if @test_server_id >= 0
+        break
+      else
+        sleep 5               # if no server is available, sleep for 5 second before it checks again
+      end
+    end
+
+    result, status = launch_test(@test_server_id, @assignment, @repo_dir, call_on)
+
+    if !status
+      #for debugging any errors in launch_test
+      server_id = @test_server_id
+      assignment = @assignment
+      repo_dir = @repo_dir
+      m_logger = MarkusLogger.instance
+
+
+      src_dir = File.join(repo_dir, assignment.repository_folder)
+
+      # Get test_dir
+      test_dir = File.join(MarkusConfigurator.markus_config_automated_tests_repository, assignment.repository_folder)
+
+      # Get the name of the test server
+      server = @list_of_servers[server_id]
+
+      # Get the directory and name of the test runner script
+      test_runner = MarkusConfigurator.markus_ate_test_runner_script_name
+
+      # Get the test run directory of the files
+      run_dir = MarkusConfigurator.markus_ate_test_run_directory
+
+
+      m_logger.log("error with launching test, result: #{result} and status: #{status}\n src_dir: #{src_dir}\ntest_dir: #{test_dir}\nserver: #{server}\ntest_runner: #{test_runner}\nrun_dir: #{run_dir}",MarkusLogger::ERROR)
+
+      # TODO: handle this error better
+      raise "error"
+    else
+      process_result(result, grouping_id, @assignment.id)
+    end
+
+  end
+
+  # From a list of test servers, choose the next available server
+  # using round-robin. Return the id of the server, and return -1
+  # if no server is available.
+  # TODO: keep track of the max num of tests running on a server
+  def self.choose_test_server()
+
+    if (defined? @last_server) && MarkusConfigurator.automated_testing_engine_on?
+      # find the index of the last server, and return the next index
+      @last_server = (@last_server + 1) % MarkusConfigurator.markus_ate_num_test_servers
+    else
+      @last_server = 0
+    end
+
+    return @last_server
+  end
+
+
+  # Launch the test on the test server by scp files to the server
+  # and run the script.
+  # This function returns two values: first one is the output from
+  # stdout or stderr, depending on whether the execution passed or
+  # had error; the second one is a boolean variable, true => execution
+  # passeed, false => error occurred.
+  def self.launch_test(server_id, assignment, repo_dir, call_on)
```

Comment:
You have my blessing to rewrite this method. Do three things:

1. Copy all the necessary files into a folder. You can pick the folder name.
2. Run the tests (I don't care if it's just a single test file) and redirect stdout and stderr to two different files.
3. Return the contents of the files, separately.